exotations/solvers/exotica_ddp_solver/include/exotica_ddp_solver/abstract_ddp_solver.h

```diff
@@ -0,0 +1,217 @@
+            HIGHLIGHT_NAMED("DDPSolver", "Final state: " << prob_->get_X(T - 1).transpose());
+        }
+
+        // copy solutions for next iteration
+        if (iteration == 0 || global_best_cost > current_cost)
+        {
+            global_best_cost = current_cost;
+            last_best_iteration = iteration;
+            global_best_U = new_U;
+        }
+
+        if (iteration - last_best_iteration > base_parameters_.FunctionTolerancePatience)
+        {
+            if (debug_) HIGHLIGHT_NAMED("DDPSolver", "Early stopping criterion reached.");
+            break;
+        }
+
+        if (last_cost - current_cost < base_parameters_.FunctionTolerance && last_cost - current_cost > 0)
+        {
+            if (debug_) HIGHLIGHT_NAMED("DDPSolver", "Function tolerance reached.");
+            break;
+        }
+
+        if (iteration == GetNumberOfMaxIterations() - 1 && debug_) HIGHLIGHT_NAMED("DDPSolver", "Max iterations reached.");
+
+        last_cost = current_cost;
+        for (int t = 0; t < T - 1; ++t)
+            prob_->Update(new_U.col(t), t);
+        prob_->SetCostEvolution(iteration, current_cost);
+    }
+
+    // store the best solution found over all iterations
+    for (int t = 0; t < T - 1; ++t)
+    {
+        solution.row(t) = global_best_U.col(t);
+        prob_->Update(global_best_U.col(t), t);
+    }
+}
+
+template <typename Initializer>
+void AbstractDDPSolver<Initializer>::SpecifyProblem(PlanningProblemPtr pointer)
+{
+    if (pointer->type() != "exotica::DynamicTimeIndexedShootingProblem")
+    {
+        ThrowNamed("This DDPSolver can't solve problem of type '" << pointer->type() << "'!");
+    }
+    MotionSolver::SpecifyProblem(pointer);
+    prob_ = std::static_pointer_cast<DynamicTimeIndexedShootingProblem>(pointer);
+    dynamics_solver_ = prob_->GetScene()->GetDynamicsSolver();
+    if (debug_) HIGHLIGHT_NAMED("DDPSolver", "initialized");
+}
+
+template <typename Initializer>
+double AbstractDDPSolver<Initializer>::ForwardPass(double alpha, Eigen::MatrixXdRef ref_trajectory)
+{
+    double cost = 0;
+    const int T = prob_->get_T();
+    const Eigen::VectorXd control_limits = dynamics_solver_->get_control_limits();
+
+    DynamicsSolverPtr dynamics_solver_ = prob_->GetScene()->GetDynamicsSolver();
+    for (int t = 0; t < T - 1; ++t)
+    {
+        Eigen::VectorXd u = prob_->get_U(t);
+        // eq. 12
+        Eigen::VectorXd delta_uk = k_gains_[t] + K_gains_[t] * dynamics_solver_->StateDelta(prob_->get_X(t), ref_trajectory.col(t));
+        u.noalias() += alpha * delta_uk;
+        // clamp controls
+        if (control_limits.size() == u.size())
+            u = u.cwiseMax(-control_limits).cwiseMin(control_limits);
+        else if (control_limits.size() == 1)
+            u = u.unaryExpr([&control_limits](double x) -> double {
+                return std::min(std::max(x, -control_limits(0)), control_limits(0));
+            });
```

Comment:
It may be better to require `control_limits` to be the same size as `u`. Ideally directly in the dynamics solver. Same for joint and velocity limits if they are used. Then this case won't be required.

Reply:
That was the original idea. However, on robots with a lot of joints this makes the XML a bit unreadable.

Reply:
When you get the initializer: If size == 1, set the same limit for all joints (i.e. setConstant() for the correct size in Eigen). If size == N, set as in the initializer. E.g. have a look at how joint velocity limits in the TimeIndexedProblems can be either 1 or N dimensions.